### ISSUE TITLE
[YCM] RuntimeError File already being parsed.

### ISSUE THREAD

**Issue by tberghammer** [Qusic/atom-youcompleteme#30]
The following error sometime displayed when saving a large file after a very short editing session (few seconds):

    [YCM] RuntimeError File already being parsed.
    Traceback (most recent call last):
      File "/home/tamas/.atom/packages/you-complete-me/ycmd/third_party/bottle/bottle.py", line 861, in _handle
        return route.call(**args)
      File "/home/tamas/.atom/packages/you-complete-me/ycmd/third_party/bottle/bottle.py", line 1734, in wrapper
        rv = callback(*a, **ka)
      File "/home/tamas/atom/atom-youcompleteme/ycmd/ycmd/../ycmd/watchdog_plugin.py", line 100, in wrapper
        return callback( *args, **kwargs )
      File "/home/tamas/atom/atom-youcompleteme/ycmd/ycmd/../ycmd/hmac_plugin.py", line 62, in wrapper
        body = callback( *args, **kwargs )
      File "/home/tamas/atom/atom-youcompleteme/ycmd/ycmd/../ycmd/handlers.py", line 70, in EventNotification
        event_handler )( request_data )
      File "/home/tamas/atom/atom-youcompleteme/ycmd/ycmd/../ycmd/completers/cpp/clang_completer.py", line 292, in OnFileReadyToParse
        with self._files_being_compiled.GetExclusive( filename ):
      File "/home/tamas/atom/atom-youcompleteme/ycmd/ycmd/../ycmd/completers/cpp/ephemeral_values_set.py", line 56, in __enter__
        raise RuntimeError( ALREADY_PARSING_MESSAGE )
    RuntimeError: File already being parsed.

The issue is caused by the fact that ycmd will return this error if we send a request to parse a file while it already parsing it. In this case it starts to parse the file for code competition and if we hit save before that parse could finish, it will return the quoted error when we send it a lint request on save.

We should find a way to ensure that only 1 request is sent to ycmd for the same file at any time (difficult because the 2 request sent by 2 function called by different providers) or add a logic to the request class to retry the request in case of a "File already being parsed." error after some delay.

**Comment by Qusic** [comment]
I am not sure how the vim plugin avoids sending duplicate parse requests. Is it safe to just ignore such kind of errors?

**Comment by tberghammer** [comment]
I checked the source code of the vim plugin and it just ignores these messages (https://github.com/Valloric/YouCompleteMe/blob/33621d409bfefabdd9c0082c374657b99cc68af8/python/ycm/client/base_request.py#L180).

In general I am not sure it is a perfect solution because in atom, I wouldn't expect the linter to try to ask for compilation errors again after a failed request what mean that we will end up in a situation where the compile errors aren't shown on save if this issue happens.

I think we have to add a retry logic to the plugin to try again the query after a few seconds if we got back a "File already being parsed" message.
